addons/website/static/src/components/website_loader/website_loader.js

```diff
@@ -1,29 +1,176 @@
+                    window.removeEventListener("beforeunload", this.showRefreshConfirmation);
+                }
+
+                return () => window.removeEventListener("beforeunload", this.showRefreshConfirmation);
+            },
+            () => [this.state.isVisible]
+        );
+
         this.props.bus.on('SHOW-WEBSITE-LOADER', this, (props) => {
             this.state.isVisible = true;
             this.state.title = props && props.title;
             this.state.showTips = props && props.showTips;
+            this.state.selectedFeatures = props && props.selectedFeatures;
+            this.state.showWaitingMessages = props && props.showWaitingMessages;
         });
         this.props.bus.on('HIDE-WEBSITE-LOADER', this, () => {
             for (const key of Object.keys(initialState)) {
                 this.state[key] = initialState[key];
             }
+            clearInterval(messagesInterval);
+            clearTimeout(this.trackModulesTimeout);
+        });
+        // Action needed if the app automatically refreshes or redirects the
+        // page without hiding/removing the WebsiteLoader. This should be
+        // called prior to any refresh/redirect if the loader is still visible.
+        this.props.bus.on("PREPARE-OUT-WEBSITE-LOADER", this, () => {
+            window.removeEventListener("beforeunload", this.showRefreshConfirmation);
         });
     }
+
+    trackModules = async (selectedFeatures) => {
+        const installInfo = await this.rpc(
+            "/website/track_installing_modules",
+            {
+                selected_features: selectedFeatures,
+                total_features: this.featuresInstallInfo.total,
+            },
+            { silent: true }
+        );
+        if (!this.featuresInstallInfo.total
+            || this.featuresInstallInfo.nbInstalled !== installInfo.nbInstalled) {
+            Object.assign(
+                this.featuresInstallInfo,
+                { nbInstalled: installInfo.nbInstalled, total: installInfo.total }
+            );
+        }
+        if (this.featuresInstallInfo.nbInstalled !== this.featuresInstallInfo.total) {
+            this.trackModulesTimeout = setTimeout(() => this.trackModules(selectedFeatures), 1000);
+        }
+    };
+
+    getWaitingMessages = (selectedFeatures) => {
+        const filteredIds = [...selectedFeatures, "last"];
+        const messagesList = this.websiteFeaturesMessages
+            .filter((msg) => filteredIds.includes(msg.id));
+        return messagesList;
+    };
+
+    showRefreshConfirmation = (ev) => {
+        if (this.state.isVisible) {
+            ev.preventDefault();
+            ev.returnValue = "";
+            return ev.returnValue;
+        }
+    };
+
+    highlightInTitle = (title, name) => {
+        const highlight = sprintf("<span class='o_website_loader_text-highlight'>%s</span>", name);
+        return markup(sprintf(title, highlight));
+    };
+
+    websiteFeaturesMessages = [{
```

Comment:
It's a class field, not a variable, so just like the functions are declared as `fctName() {}` and not `function fctName() {}`, this doesn't need `const` (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#field_declarations).
I guess I could extract it as a constant outside of the component instead. Do you think there should be a reason to do so?

Reply:
Okay indeed :open_mouth:
> Do you think there should be a reason to do so

Not really :thinking: Since it works, it can stay like this I guess :wink: